src/sleep.rs

```diff
@@ -0,0 +1,110 @@
+
+#[derive(Debug)]
+pub enum WakeupSource {
+    Timer { dur: Duration },
+    Uart { uart_num: i32, threshold: i32 },
+    // TODO: add more sources
+}
+
+impl WakeupSource {
+    pub fn apply(&self) -> Result<(), EspError> {
+        match *self {
+            WakeupSource::Timer { dur } => {
+                esp!(unsafe { esp_sleep_enable_timer_wakeup(dur.as_micros() as u64) })?;
+            }
+            WakeupSource::Uart {
+                uart_num,
+                threshold,
+            } => {
+                esp!(unsafe { uart_set_wakeup_threshold(uart_num, threshold) })?;
+                esp!(unsafe { esp_sleep_enable_uart_wakeup(uart_num) })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub trait Sleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError>;
+    fn sleep(&self) -> Result<(), EspError>;
+
+    fn apply_wakeup_sources(&self, wakeup_sources: &[WakeupSource]) -> Result<(), EspError> {
+        // reset wakeup sources
+        esp!(unsafe { esp_sleep_disable_wakeup_source(esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL) })?;
+
+        for wakeup in wakeup_sources.iter() {
+            wakeup.apply()?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct LightSleep {
+    wakeup_sources: Vec<WakeupSource>,
+}
+
+impl LightSleep {
+    pub fn timer(dur: Duration) -> Result<Self, EspError> {
+        // TODO: checks for duplicate wakeup sources
+
+        let mut lsleep = Self::default();
+        lsleep.add_wakeup_source(WakeupSource::Timer { dur })?;
+        Ok(lsleep)
+    }
+}
+
+impl Sleep for LightSleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError> {
+        self.wakeup_sources.push(wakeup);
+        Ok(())
+    }
+
+    fn sleep(&self) -> Result<(), EspError> {
+        self.apply_wakeup_sources(&self.wakeup_sources)?;
+
+        esp!(unsafe { esp_light_sleep_start() })?;
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct DeepSleep {
+    wakeup_sources: Vec<WakeupSource>,
+}
+
+impl DeepSleep {
+    pub fn timer(dur: Duration) -> Result<Self, EspError> {
+        let mut dsleep = Self::default();
+        dsleep.add_wakeup_source(WakeupSource::Timer { dur })?;
+        Ok(dsleep)
+    }
+}
+
+impl Sleep for DeepSleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError> {
+        // TODO: checks for duplicate wakeup sources
+
+        if matches!(wakeup, WakeupSource::Uart { .. }) {
+            return Err(EspError::from_infallible::<ESP_ERR_INVALID_ARG>());
+        }
+        self.wakeup_sources.push(wakeup);
+        Ok(())
+    }
+
+    fn sleep(&self) -> Result<(), EspError> {
+        self.apply_wakeup_sources(&self.wakeup_sources)?;
+
+        unsafe { esp_deep_sleep_start() };
+        #[allow(unreachable_code)]
+        // if function returns, it means that deep sleep has been rejected
+        Err(EspError::from_infallible::<ESP_FAIL>())
```

Comment:
Hmmm. This would prevent our `deep_sleep` function from having the `-> !` signature. :(
Shall we `panic!` here instead, if the `esp_deep_sleep_start` function fails to run? Not sure myself, but might be the better option, given that we check enough preconditions before entering deep sleep, to make sure that panicing would be next to impossible - as in checking for a too short sleep interval.

Reply:
yeah, this is abit confusing, I would say that `-> !` would make alot of sense here, but didnt want to have that since the documentation says that `esp_deep_sleep_start`might actually return. 

Seems to me that the only reason that would happen is that the timer duration is too short, but i was not able to find out what "too short" means. So not sure we can check that?

My reasoning here was that you actually can "try" to deep sleep, but if that doesnt work you can actually check that and do whatever, try again perhaps? Doing a panic here doesnt "feel right", since you can end up doing something different that you asked for.

But then again... maybe panic is a good solution, it will make the function signature more self explanatory. If the only way to get to that panic is to ask for a too short deep sleep duration. And the functional difference between a panic and a very very short deep sleep is not much i guess.